client/finality-grandpa/src/finality_proof.rs

```diff
@@ -305,7 +305,9 @@ pub fn prove_warp_sync<Block: BlockT, B: BlockchainBackend<Block>>(
 		index = index + One::one();
 	}
 
-	if result.last().as_ref().map(|head| head.header.number()) != Some(&end_number) {
+	let at_limit = max_fragment_limit.map(|limit| result.len() >= limit).unwrap_or(false);
+
+	if !at_limit && result.last().as_ref().map(|head| head.header.number()) != Some(&end_number) {
```

Comment:
:+1: when at_limit we probably had twice the fragment
Actually realize this condition was not really simple to read, the point was just add the last finalized block if not already added, so I added a code suggestion bellow.